src/main/java/com/aws/greengrass/dependency/Context.java

```diff
@@ -553,13 +563,17 @@ private Constructor<T> pickConstructor(Class<T> clazz) throws NoSuchMethodExcept
          * @return the current (existing or computed) object instance
          * @throws E when mapping function throws checked exception
          */
-        public final synchronized <E extends Exception> T computeObjectIfEmpty(
+        public final <E extends Exception> T computeObjectIfEmpty(
                 CrashableFunction<Value, T, E> mappingFunction) throws E {
-            if (object != null) {
-                return object;
-            }
+            synchronized (Context.this) {
+                synchronized (this) {
+                    if (object != null) {
+                        return object;
+                }
```

Comment:
you need to keep holding the lock when calling put and inject, or else the semantics of this change do not match the previous implementation

Reply:
Oops, IDE had this collapsed so the indent is wrong. Will fix. The putAndInject is included in the lock scope